### ISSUE TITLE
[GroheOndus] Grohe Sense Thing missing channels

### ISSUE THREAD

**Issue by vzorglub** [openhab/openhab2-addons#3898]
## Expected Behavior
The Grohe Sense reports Temperature, Battery, Humidity and Water/Flood
These channels are not available. The channels available are for the Grohe Sense Gard

## Possible Solution
Please add the channels for Battery, Humidity and Water/Flood for the Grohe Sense (Battery powered)
The battery channel is not required for the Grohe Sense+ (Mains powered)

@FlorianSW , Would that be possible, I am happy to send you the device for testing if required

**Comment by vzorglub** [comment]
Also the temperature channel is throwing an error:

```
2018-09-09 09:39:54.468 [.ItemChannelLinkAddedEvent] - Link 'Toilet_ThermostatAmbientTemp-groheondus:appliance:c4a2ccb4-0906-4470-b021-c8ee0ace85e7:temperature' has been added.
==> /var/log/openhab2/openhab.log <==
2018-09-09 09:39:55.146 [ERROR] [ome.core.thing.link.ThingLinkManager] - Exception occurred while informing handler: null

java.lang.NullPointerException: null
	at org.eclipse.smarthome.core.library.types.QuantityType.<init>(QuantityType.java:104) ~[?:?]
	at org.openhab.binding.groheondus.internal.handler.GroheOndusApplianceHandler.updateChannel(GroheOndusApplianceHandler.java:134) ~[?:?]
	at org.openhab.binding.groheondus.internal.handler.GroheOndusApplianceHandler.channelLinked(GroheOndusApplianceHandler.java:95) ~[?:?]
	at org.eclipse.smarthome.core.thing.link.ThingLinkManager.lambda$0(ThingLinkManager.java:300) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
```

**Comment by FlorianSW** [comment]
Hi,

thanks for opening the issue :) I extracted the teamperature NPE to another issue (see #3900), as well as the thing initialization error you mentioned in the PR #3696 to #3899.

It's totally possible, that the device model returned from the API is different from the Grohe Sense guard device, which would explain these errors. However, in the account I used to develop this binding, only the sense guard was available, so in order to get this thing developed correctly, I would need an GROHE account, which has at least one SENSE device available (no need to send the actual device to someone).

If you would be ok with sharing your account with me as long as the thing is developed (it could take some days based on my availability :/) that would be awesome! In this case, could you change the password of your account to something new randomly created, safe password and send it to me via e-mail: florian.schmidt.stargatewissen@gmail.com

Once the thing is finished, you should change the password back to what it was before, or even better to something newly secure and randomly created one :)

**Comment by vzorglub** [comment]
```
as long as the thing is developed
```
How do I send you the thing?
I'll be happy to send you the account details

**Comment by FlorianSW** [comment]
> How do I send you the thing?
No need to send me the actual device, I only need the data from the API, which I would be able to access whith the account credentials :)

> I'll be happy to send you the account details
Awesome! Thanks :)

**Comment by vzorglub** [comment]
I can't find the password settings in the android app so I have sent an email to Grohe
This is one of the worst designed app I have ever seen and it's almost 100MB!!
What's wrong with them??

**Comment by FlorianSW** [comment]
> I can't find the password settings in the android app

On the main screen (before selecting a house or a device) there's a hamburger menu at the bottom left (an icon with 3 lines). Pressing it opens a popup where you should see "settings". There you can (at the very top) select "Change password". That's at least at the latest version I've installed :P

> This is one of the worst designed app I have ever seen and it's almost 100MB!!

Apart from that, I need to say, that the API GROHE designed there is very good from my point of view. They're strictly following REST principals, so that without knowing much about their implementation, it was pretty easy to navigate through the api and getting the different information, very positive! I would love to have more of these APIs and I'm wondering why they're not officially making this API public and documented somewhere how to talk to it. There's nothing they need to fear about, I think, it's a nice api :) (This, however, is just my technical point of view, think about why this binding exists, not because this app is very intuitive :P)

**Comment by vzorglub** [comment]
Never mind I found it. Not very intuitive...
I have sent you the email with the details after changing the password.

**Comment by vzorglub** [comment]
Thanks!!! That was fast and it works for the temperature and humidity channels
However, this sensor is designed as a flood sensor, generally to be placed on the floor to detect leaks and floods.
Could you add that channel, please.
If you want me to trigger it, we will need to synchronize somehow.
Also did you find a battery channel?

**Comment by FlorianSW** [comment]
> However, this sensor is designed as a flood sensor, generally to be placed on the floor to detect leaks and floods.

So, the sense sends you a notification once a leak is detected, right? This is a bit problematic (the SENSE guard as a similar functionality with the additional "turn the water supply off" feature). The problem is, that the notification interface is Firebase Cloud Messaging, which is a background service in the android app (there's a similar service for iOS). There's no way (or at least none I know) to get this working in a pure java only server application, as there needs to be a permanent connection to a firebase server in order to "receive" notifications.

However, the API exposes a notifications endpoint, which would look like:
https://idp-apigw.cloud.grohe.com/v2/iot/locations/{locationId}/rooms/[roomId}/appliances/{applianceId}/notifications
When called with an appropriate set of credentials it will return something like that:
```
[
    {
        "id": "notificationId",
        "category": 20,
        "is_read": true,
        "timestamp": "notificationDate",
        "type": 31,
        "threshold_quantity": "no",
        "threshold_type": "no"
    }
]
```

So, there're some problems:
* I've no idea what the types and catgeories means, however there are most likely some references somewhere in the de-compiled app code
* This is a pull, instead of a push, that means, that openHAB would need to issue a request to this endpoint to see if there're new notifications

I could think about a shorter-intervalled request (say, 1 minute) which is for this endpoint only (all other data will still be pulled every 15 minutes by default) requesting the last notification. If there's a change, there could be some channels (like last_notification_timestamp, last_notifiation_category and last_notification_type) which get's changed once the 1-minute interval request gets a new notification. Then, openHAB rules can be used to "trigger" on this change and, e.g., send a broadcast notification to registered android and iOS devices (when openHAB-cloud is setup). However, I'm not sure, what performance implications that has when one binding issues one request every 30 sedonds or even 1 minute (to get as close to the leak time as possible), @kaikreuzer can you say something about that? :)

> Also did you find a battery channel?

No, is there some data in the app?

**Comment by martinvw** [comment]
Note that the nest binding also uses a fire base implementation, I suppose
@wborn could help you out if needed
Op zo 9 sep. 2018 om 20:07 schreef Florian <notifications@github.com>

**Comment by FlorianSW** [comment]
@martinvw Great, thanks for the info :) I'll take a look into it and see how this can be done here, too :)

> Also did you find a battery channel?

Looking into the api again, there's actually an endpoint which returns some status information, I'm working on getting that into the PR, too :)

**Comment by wborn** [comment]
> Note that the nest binding also uses a fire base implementation

Nest has Firebase support but it's [no longer recommended](https://developers.nest.com/guides/samples/firebase-sdks).

The Nest binding doesn't use Firebase it uses:
* [Nest REST API](https://developers.nest.com/documentation/api-reference) for commands 
* [REST Streaming](https://developers.nest.com/guides/api/rest-streaming-guide) (using Server-Sent Events) for updates

**Comment by vzorglub** [comment]
Maybe the flooded state is hidden in there

**Comment by martinvw** [comment]
I thought that the stream of data was firebase, thanks for correcting me @wborn

**Comment by FlorianSW** [comment]
> Maybe the flooded state is hidden in there

Most likely not. The only types the app recognizes from this data is:
* battery
* connection (whatever that means)
* update_available

**Comment by vzorglub** [comment]
Meh......

**Comment by FlorianSW** [comment]
Can you show me the configuration of your things (with anonymized username, password and IDs)?

**Comment by vzorglub** [comment]
Hi, Florian, what do you mean exactly
I have configured in the PaperUI
![image](https://user-images.githubusercontent.com/19504096/45407774-5a219280-b662-11e8-98ee-d3ed558617d5.png)

![image](https://user-images.githubusercontent.com/19504096/45407823-78878e00-b662-11e8-8245-dbaeb5bc7ccf.png)


![image](https://user-images.githubusercontent.com/19504096/45407845-86d5aa00-b662-11e8-9bf2-b43c4060d582.png)

**Comment by FlorianSW** [comment]
Oops, my fault, sorry :/ Should be fixed in the latest commit (give it some time to be built by the CI).

**Comment by vzorglub** [comment]
Hello Florian,
Did that commit ever got built?

**Comment by FlorianSW** [comment]
Should've been built, yes, see: https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/binding/org.openhab.binding.groheondus/2.4.0-SNAPSHOT/

**Comment by FlorianSW** [comment]
Ok, I pushed some new commits, let's see if these are getting built correctly :) Sorry for the inconvenience!

**Comment by vzorglub** [comment]
Today's build:

openhab.log:

```
2018-09-30 21:44:42.592 [ERROR] [nal.common.AbstractInvocationHandler] - An error occurred while calling method 'ThingHandler.initialize()' on 'org.openhab.binding.groheondus.internal.handler.GroheOndusApplianceHandler@bfeef81': null
java.lang.NullPointerException: null
	at org.openhab.binding.groheondus.internal.handler.GroheOndusApplianceHandler.getPollingInterval(GroheOndusApplianceHandler.java:120) ~[?:?]
	at org.openhab.binding.groheondus.internal.handler.GroheOndusApplianceHandler.initialize(GroheOndusApplianceHandler.java:75) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
	at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:153) [101:org.eclipse.smarthome.core:0.10.0.201809271800]
	at org.eclipse.smarthome.core.internal.common.Invocation.call(Invocation.java:53) [101:org.eclipse.smarthome.core:0.10.0.201809271800]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
2018-09-30 21:44:42.599 [ERROR] [ome.core.thing.internal.ThingManager] - Exception occurred while initializing handler of thing 'groheondus:appliance:c4a2ccb4-0906-4470-b021-c8ee0ace85e7': null
java.lang.NullPointerException: null
	at org.openhab.binding.groheondus.internal.handler.GroheOndusApplianceHandler.getPollingInterval(GroheOndusApplianceHandler.java:120) ~[?:?]
	at org.openhab.binding.groheondus.internal.handler.GroheOndusApplianceHandler.initialize(GroheOndusApplianceHandler.java:75) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
	at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:153) [101:org.eclipse.smarthome.core:0.10.0.201809271800]
	at org.eclipse.smarthome.core.internal.common.Invocation.call(Invocation.java:53) [101:org.eclipse.smarthome.core:0.10.0.201809271800]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
2018-09-30 21:44:42.602 [hingStatusInfoChangedEvent] - 'groheondus:appliance:c4a2ccb4-0906-4470-b021-c8ee0ace85e7' changed from INITIALIZING to UNINITIALIZED (HANDLER_INITIALIZING_ERROR)
```

**Comment by vzorglub** [comment]
Hi @FlorianSW 
Updated to M6 and the binding is not recognising the SENSE device.
NO channel for humidity and errors in log:

![image](https://user-images.githubusercontent.com/19504096/48919410-8f303980-ee8a-11e8-8e2b-41b207099095.png)

Logs:
```
2018-11-22 19:11:15.054 [ERROR] [nal.common.AbstractInvocationHandler] - An error occurred while calling method 'ThingHandler.initialize()' on 'org.openhab.binding.groheondus.internal.handler.GroheOndusApplianceHandler@59bf11fd': null

java.lang.NullPointerException: null
	at org.openhab.binding.groheondus.internal.handler.GroheOndusApplianceHandler.getPollingInterval(GroheOndusApplianceHandler.java:120) ~[?:?]
	at org.openhab.binding.groheondus.internal.handler.GroheOndusApplianceHandler.initialize(GroheOndusApplianceHandler.java:75) ~[?:?]
	at sun.reflect.GeneratedMethodAccessor84.invoke(Unknown Source) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
	at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:153) [101:org.eclipse.smarthome.core:0.10.0.201811171951]
	at org.eclipse.smarthome.core.internal.common.Invocation.call(Invocation.java:53) [101:org.eclipse.smarthome.core:0.10.0.201811171951]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]

2018-11-22 19:11:15.067 [ERROR] [core.thing.internal.ThingManagerImpl] - Exception occurred while initializing handler of thing 'groheondus:appliance:c4a2ccb4-0906-4470-b021-c8ee0ace85e7': null

java.lang.NullPointerException: null
	at org.openhab.binding.groheondus.internal.handler.GroheOndusApplianceHandler.getPollingInterval(GroheOndusApplianceHandler.java:120) ~[?:?]
	at org.openhab.binding.groheondus.internal.handler.GroheOndusApplianceHandler.initialize(GroheOndusApplianceHandler.java:75) ~[?:?]
	at sun.reflect.GeneratedMethodAccessor84.invoke(Unknown Source) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
	at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:153) [101:org.eclipse.smarthome.core:0.10.0.201811171951]
	at org.eclipse.smarthome.core.internal.common.Invocation.call(Invocation.java:53) [101:org.eclipse.smarthome.core:0.10.0.201811171951]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
```

**Comment by FlorianSW** [comment]
Hmm, it seems you do not have the current version of the binding installed, as the class `GroheOndusApplianceHandler` does not exist in the latest version of the pull request anymore. Could you update it to the latest build version?

**Comment by vzorglub** [comment]
Hi @FlorianSW 
Any update on this issue, please?

**Comment by vzorglub** [comment]
Ok, restarted OH.
Really clean restart sequence except the Grohe binding!!
Error:
```
2018-11-29 07:41:41.762 [ERROR] [nal.common.AbstractInvocationHandler] - An error occurred while calling method 'ThingHandler.initialize()' on 'org.openhab.binding.groheondus.internal.handler.GroheOndusApplianceHandler@1614c142': null
java.lang.NullPointerException: null
	at org.openhab.binding.groheondus.internal.handler.GroheOndusApplianceHandler.getPollingInterval(GroheOndusApplianceHandler.java:120) ~[?:?]
	at org.openhab.binding.groheondus.internal.handler.GroheOndusApplianceHandler.initialize(GroheOndusApplianceHandler.java:75) ~[?:?]
	at sun.reflect.GeneratedMethodAccessor81.invoke(Unknown Source) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
	at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:153) [101:org.eclipse.smarthome.core:0.10.0.201811281422]
	at org.eclipse.smarthome.core.internal.common.Invocation.call(Invocation.java:53) [101:org.eclipse.smarthome.core:0.10.0.201811281422]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
```

The same error occrus 4 times in a row whilst the system tries to initialise the thing
